### PR TITLE
WB-1458: Add `autoFocus` and `enableTypeAhead` props to `SingleSelect`.

### PR DESCRIPTION
**Pull request by jandrade**
## Summary:

Some users have reported some keyboard issues with the `TimePicker`
component created in webapp. This component is used to be able to either
select a time slot or type in a specific time.

This PR adds a new prop called `autoFocus` to the `SingleSelect`
component so we can prevent the input focus from being stolen by the
listbox.

Also, this PR adds a new prop called `enableTypeAhead` to the
`SingleSelect` component to avoid the listbox from being automatically
focused when the user types in the input.

Issue: WB-1458

## Test Plan:

Navigate to the `AutoFocus` story (`/?path=/story/dropdown-singleselect--auto-focus-disabled`) and verify that the input works as
expected. This means that if the user focuses on the input and/or if the
user stars typing, the listbox should not be focused.

BEFORE

https://user-images.githubusercontent.com/843075/204825294-d18b941e-40b6-4665-92b3-90a77fc21bff.mov

AFTER

https://user-images.githubusercontent.com/843075/204825322-3003ce72-2322-48e2-a5cb-484c245469bd.mov

### CODE REVIEW

File: packages/wonder-blocks-dropdown/src/components/__docs__/single-select.argtypes.js [hunk 1]
```diff
@@ -1,54 +0,0 @@
-// @flow
```
Comment (jandrade) [inline packages/wonder-blocks-dropdown/src/components/__docs__/single-select.argtypes.js:1]: **note:** I noticed that this file was not being used.

File: packages/wonder-blocks-dropdown/src/components/dropdown-core.js [hunk 2]
```diff
@@ -106,24 +126,13 @@ type ItemAriaRole = "option" | "menuitem";
...
+    // Required props
 
     /**
-     * An optional string that the user entered to search the items. When this
-     * and the onSearchTextChanged exist, SearchField will be displayed at the
-     * top of the dropdown body.
+     * Items for the menu.
      */
-    searchText?: ?string,
+    items: Array<DropdownItem>,
```
Comment (jandrade) [inline packages/wonder-blocks-dropdown/src/components/dropdown-core.js:135]: **note:** Decided to do a minor refactor to sort props alphabetically.

File: packages/wonder-blocks-dropdown/src/components/single-select.js [hunk 3]
```diff
@@ -157,13 +179,6 @@ type State = {|
     openerElement: ?HTMLElement,
 |};
 
-type DefaultProps = {|
```
Comment (jandrade) [inline packages/wonder-blocks-dropdown/src/components/single-select.js:160]: **note:** Found an issue with Storybook + [Indexed Access Types](https://flow.org/en/docs/types/indexed-access/), so I had to move the default props above.

File: packages/wonder-blocks-dropdown/src/components/dropdown-core.js [hunk 4]
```diff
@@ -72,18 +72,38 @@ type Labels = {|
...
     /**
      * Whether this menu should be left-aligned or right-aligned with the
      * opener component. Defaults to left-aligned.
      */
     alignment: "left" | "right",
 
+    /**
+     * Whether to auto focus an option.
+     */
+    autoFocus: boolean,
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/src/components/dropdown-core.js:84]: **suggestion (non-blocking):** Could you include what the default value is in the docstring as well as when people my want to use the non-default value?

File: packages/wonder-blocks-dropdown/src/components/dropdown-core.js [hunk 5]
```diff
@@ -72,18 +72,38 @@ type Labels = {|
...
+     * - Type a character: focus moves to the next item with a name that starts
+     *   with the typed character.
+     * - Type multiple characters in rapid succession: focus moves to the next
+     *   item with a name that starts with the string of characters typed.
+     *
+     * **NOTE:** Type-ahead is recommended for all listboxes, but there might be
+     * some cases where it's not desirable (for example when using a `TextField`
+     * as the opener element).
+     */
+    enableTypeAhead: boolean,
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/src/components/dropdown-core.js:99]: **suggestion (non-blocking):** Could you include what the default value for this is in the docstring?

File: packages/wonder-blocks-dropdown/src/components/dropdown-core.js [hunk 6]
```diff
@@ -397,7 +424,11 @@ class DropdownCore extends React.Component<Props, State> {
     // Figure out focus states for the dropdown after it has changed from open
     // to closed or vice versa
     initialFocusItem() {
-        const {open} = this.props;
+        const {autoFocus, open} = this.props;
+
+        if (!autoFocus) {
+            return;
+        }
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/src/components/dropdown-core.js:431]: **suggestion (non-blocking):** Now that we aren't always focusing an item, this should maybe be renamed to `maybeFocusInitialItem()`.

File: packages/wonder-blocks-dropdown/src/components/single-select.js [hunk 7]
```diff
@@ -39,8 +39,53 @@ export type SingleSelectLabels = {|
...
+     * - Type a character: focus moves to the next item with a name that starts
+     *   with the typed character.
+     * - Type multiple characters in rapid succession: focus moves to the next
+     *   item with a name that starts with the string of characters typed.
+     *
+     * **NOTE:** Type-ahead is recommended for all listboxes, but there might be
+     * some cases where it's not desirable (for example when using a `TextField`
+     * as the opener element).
+     */
+    enableTypeAhead: boolean,
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/src/components/single-select.js:67]: **suggestion (non-blocking):** Include what the default values for these props are in the docstrings.

File: packages/wonder-blocks-dropdown/src/components/single-select.js [hunk 8]
```diff
@@ -39,8 +39,53 @@ export type SingleSelectLabels = {|
...
+
+    /**
+     * Whether to display the "light" version of this component instead, for
+     * use when the component is used on a dark background.
+     */
+    light: boolean,
+    /**
+     * The object containing the custom labels used inside this component.
+     */
+    labels: SingleSelectLabels,
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/src/components/single-select.js:83]: **style nit (non-blocking):** missing space between items.

File: packages/wonder-blocks-dropdown/src/components/__docs__/single-select.stories.js [hunk 9]
```diff
@@ -516,3 +528,90 @@ CustomLabels.parameters = {
...
+                            size="small"
+                            style={styles.icon}
+                        />
+                    </View>
+                )}
+            >
+                {timeSlotOptions}
+            </SingleSelect>
+        </View>
+    );
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/src/components/__docs__/single-select.stories.js:597]: **commentary:** This looks like the start of a `ComboBox` component. 😉

**Review by kevinbarabash — APPROVED** [review]
LGTM.  These new props look super useful, especially for the creation of new shared components.  If these props are on `SingleSelect`, should they also be on `MultiSelection` and `ActionMenu`?

**Comment by jandrade** [comment]
> LGTM. These new props look super useful, especially for the creation of new shared components. If these props are on SingleSelect, should they also be on MultiSelection and ActionMenu?

Absolutely! I created a ticket to work on that at some point next month: https://khanacademy.atlassian.net/browse/WB-1471

**Pull request merged**